README.md
Centrifuge + Go = Centrifugo – harder, better, faster, stronger.

Centrifugo is a real-time messaging server. This is a successor of 
[Centrifuge](https://github.com/shilkin/centrifuge). Please note that it can be used in conjuction with your application backend written in any language - Python, Ruby, Perl, PHP, Javascript, Java, Objective-C etc.

To understand what is it for and how it works – please, read 
[documentation](http://fzambia.gitbooks.io/centrifugal/content/) of 
Centrifugal organization.

Releases available as single executable files – just 
[download](https://github.com/shilkin/centrifugo/releases) archive for your platform,  
unpack it and you are done. See also [Docker image](https://registry.hub.docker.com/u/fzambia/centrifugo/).

Try [demo instance](https://centrifugo.herokuapp.com/) on Heroku (password `demo`). Or deploy your own Centrifugo instance in one click:

[![Deploy](https://www.herokucdn.com/deploy/button.png)](https://heroku.com/deploy?template=https://github.com/shilkin/centrifugo)

Highlights:
* Fast server capable to serve lots of simultanious connections
* SHA-256 HMAC-based authorization
* HTTP API to communicate from your application backend (publish messages in channels etc.). API clients for Python, Ruby, PHP. Simple to implement new one.
* Javascript client to connect from web browser over SockJS or pure Websocket protocol
* Presence, history and join/leave events for channels
* Flexible configuration of channels via namespaces
* Different types of channels - private, user limited, client limited channels
* Administrative web interface
* Scale to several machines with Redis
* Ready to deploy (docker, rpm, Nginx configuration)
* Easily integrates with existing application - no need to rewrite your backend code
* MIT license

Simplified scheme:

![scheme](https://raw.githubusercontent.com/centrifugal/documentation/master/assets/images/scheme.png)


[![Join the chat at https://gitter.im/centrifugal/centrifugo](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/centrifugal/centrifugo?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)


debian/changelog
centrifugo (0.1.0ubuntu15) utopic; urgency=medium

  * fix history call

 -- Denis <autobuilder@corp.sputnik.ru>  Tue, 15 Nov 2016 14:20:58 +0300

centrifugo (0.1.0ubuntu14) utopic; urgency=medium

  * MAILPAAS-3148| [refactoring] user-independent channel subscription
  * MAILPAAS-3148| fix unused import
  * fix build: go-tarantool opts changed

 -- shilkin <autobuilder@corp.sputnik.ru>  Mon, 14 Nov 2016 18:31:03 +0300

centrifugo (0.1.0ubuntu13) utopic; urgency=medium

  * MAILPAAS-1841| Added notification extender
  * MAILPAAS-1841| Notification extension supported
  * MAILPAAS-1841| Extender configuration added + notification id type changed from int to string
  * MAILPAAS-1841| Renaming: make some private structures public
  * Fix CMakeLists.txt: add ttconnector-framework dependency
  * Fix debian/control: add schema-dev dependency

 -- shilkin <autobuilder@corp.sputnik.ru>  Tue, 10 May 2016 12:32:32 +0300

centrifugo (0.1.0ubuntu12) utopic; urgency=medium

  * MAILPAAS-1248| Blocking connect to each node of tarantool from pool
  * MAILPAAS-1248| Blocking connect to each node of tarantool from pool: just skip an error in case of opts.Reconnect > 0

 -- shilkin <autobuilder@corp.sputnik.ru>  Mon, 15 Feb 2016 17:41:05 +0300

centrifugo (0.1.0ubuntu11) utopic; urgency=medium

  * Additional logging during tarantool connection pool initialization

 -- shilkin <autobuilder@corp.sputnik.ru>  Thu, 05 Nov 2015 13:01:07 +0300

centrifugo (0.1.0ubuntu10) utopic; urgency=medium

  * Version bump for Jenkins repair
  * Version bump for Jenkins repair

 -- shilkin <autobuilder@corp.sputnik.ru>  Thu, 05 Nov 2015 10:21:18 +0300

centrifugo (0.1.0ubuntu9) utopic; urgency=medium

  * Jenkins bump

 -- Vitaly Isaev <autobuilder@corp.sputnik.ru>  Tue, 29 Sep 2015 11:24:26 +0300

centrifugo (0.1.0ubuntu8) utopic; urgency=medium

  * MAILPAAS-504 | Переезд с shilkin/go-tarantool на tarantool/go-tarantool.

 -- shilkin <autobuilder@corp.sputnik.ru>  Thu, 10 Sep 2015 12:12:11 +0300

centrifugo (0.1.0ubuntu7) utopic; urgency=medium

  * main.go fixed

 -- shilkin <autobuilder@corp.sputnik.ru>  Mon, 07 Sep 2015 20:22:30 +0300

centrifugo (0.1.0ubuntu6) utopic; urgency=medium

  [ FZambia ]
  * update sockjs-go dependency to its latest v2 revision
  * update viper dependency and handle ConfigPareseError
  * godeps
  * use packages from godep workspace
  * update gorilla websocket
  * update securecookie
  * v0.2.4
  * fix wrong names in comments
  * tests for admin token and ping command
  * recursively vendor dependencies
  * add missing tests for logger
  * write more tests
  * more testing fo handlers

  [ Alexandr Emelin ]
  * test for AdminWebsocketHandler

  [ FZambia ]
  * test RedisEngine, check that Redis API message project is not empty
  * more test coverage for memory engine
  * better locking for redis engine

  [ shilkin ]
  * Merge remote-tracking branch 'upstream/master' into MERGE-UPSTREAM
  * Убрана отладочная печать, скорректирован config.json
  * Убран комментарий

  [ FZambia ]

 -- FZambia <autobuilder@corp.sputnik.ru>  Mon, 07 Sep 2015 14:03:16 +0300

centrifugo (0.1.0ubuntu5) utopic; urgency=medium

  * Fix problem with casting counter to int64/uint64

 -- shilkin <autobuilder@corp.sputnik.ru>  Fri, 04 Sep 2015 17:31:57 +0300

centrifugo (0.1.0ubuntu4) utopic; urgency=medium

  * Счётчик кастится к uint64
  * Счётчик кастится к uint64

 -- shilkin <autobuilder@corp.sputnik.ru>  Fri, 28 Aug 2015 19:29:26 +0300

centrifugo (0.1.0ubuntu3) utopic; urgency=medium

  * Испоольщование порта из настроек для передачи в TarantoolEngine

 -- shilkin <autobuilder@corp.sputnik.ru>  Fri, 28 Aug 2015 18:10:37 +0300

centrifugo (0.1.0ubuntu2) utopic; urgency=medium

  * Bumping build

 -- Autobuilder <autobuilder@corp.sputnik.ru>  Fri, 28 Aug 2015 15:43:58 +0300

centrifugo (0.1.0ubuntu1) unstable; urgency=low

  * Starting build

 -- Nobody <autobuilder@corp.sputnik.ru>  Thu, 27 Aug 2015 16:22:34 +0300


main.go
package main

import (
	"fmt"
	"net/http"
	"os"
	"os/signal"
	"runtime"
	"strings"
	"syscall"
	"time"

	"github.com/shilkin/centrifugo/Godeps/_workspace/src/github.com/spf13/cobra"
	"github.com/shilkin/centrifugo/Godeps/_workspace/src/github.com/spf13/viper"
	"github.com/shilkin/centrifugo/libcentrifugo"
	"github.com/shilkin/centrifugo/libcentrifugo/extender"
	"github.com/shilkin/centrifugo/libcentrifugo/logger"
	"github.com/tarantool/go-tarantool"
)

const (
	VERSION = "0.2.4"
)

var configFile string

func setupLogging() {
	logLevel, ok := logger.LevelMatches[strings.ToUpper(viper.GetString("log_level"))]
	if !ok {
		logLevel = logger.LevelInfo
	}
	logger.SetLogThreshold(logLevel)
	logger.SetStdoutThreshold(logLevel)

	if viper.IsSet("log_file") && viper.GetString("log_file") != "" {
		logger.SetLogFile(viper.GetString("log_file"))
		// do not log into stdout when log file provided
		logger.SetStdoutThreshold(logger.LevelNone)
	}
}

func handleSignals(app *libcentrifugo.Application) {
	sigc := make(chan os.Signal, 1)
	signal.Notify(sigc, syscall.SIGHUP, syscall.SIGINT, os.Interrupt)
	for {
		sig := <-sigc
		logger.INFO.Println("Signal received:", sig)
		switch sig {
		case syscall.SIGHUP:
			// reload application configuration on SIGHUP
			logger.INFO.Println("Reloading configuration")
			err := viper.ReadInConfig()
			if err != nil {
				switch err.(type) {
				case viper.ConfigParseError:
					logger.CRITICAL.Printf("Error parsing configuration: %s\n", err)
					continue
				default:
					logger.CRITICAL.Println("Unable to locate config file")
					continue
				}
			}
			setupLogging()
			c := newConfig()
			s := structureFromConfig(nil)
			app.SetConfig(c)
			app.SetStructure(s)
			logger.INFO.Println("Configuration successfully reloaded")
		case syscall.SIGINT, os.Interrupt:
			logger.INFO.Println("Shutting down")
			go time.AfterFunc(5*time.Second, func() {
				os.Exit(1)
			})
			app.Shutdown()
			os.Exit(130)
		}
	}
}

// Main runs Centrifugo as a service.
func Main() {

	var port string
	var address string
	var debug bool
	var name string
	var web string
	var engn string
	var logLevel string
	var logFile string
	var insecure bool
	var useSSL bool
	var sslCert string
	var sslKey string

	var redisHost string
	var redisPort string
	var redisPassword string
	var redisDB string
	var redisURL string
	var redisAPI bool
	var redisPool int

	// Tarantool
	var ttHost string
	var ttPort string
	var ttUser string
	var ttPassword string
	var ttPool int
	var ttTimeoutResponse int
	var ttTimeoutReconnect int
	var ttMaxReconnect uint
	var ttAggrNode string
	var ttAggrPort string
	var ttServiceName string
	var ttACL string

	var rootCmd = &cobra.Command{
		Use:   "",
		Short: "Centrifugo",
		Long:  "Centrifuge + GO = Centrifugo – harder, better, faster, stronger",
		Run: func(cmd *cobra.Command, args []string) {

			viper.SetDefault("gomaxprocs", 0)
			viper.SetDefault("prefix", "")
			viper.SetDefault("web_password", "")
			viper.SetDefault("web_secret", "")
			viper.RegisterAlias("cookie_secret", "web_secret")
			viper.SetDefault("max_channel_length", 255)
			viper.SetDefault("channel_prefix", "centrifugo")
			viper.SetDefault("node_ping_interval", 5)
			viper.SetDefault("message_send_timeout", 60)
			viper.SetDefault("expired_connection_close_delay", 10)
			viper.SetDefault("presence_ping_interval", 25)
			viper.SetDefault("presence_expire_interval", 60)
			viper.SetDefault("private_channel_prefix", "$")
			viper.SetDefault("namespace_channel_boundary", ":")
			viper.SetDefault("user_channel_boundary", "#")
			viper.SetDefault("user_channel_separator", ",")
			viper.SetDefault("client_channel_boundary", "&")
			viper.SetDefault("sockjs_url", "https://cdn.jsdelivr.net/sockjs/1.0/sockjs.min.js")

			viper.SetDefault("project_name", "")
			viper.SetDefault("project_secret", "")
			viper.SetDefault("project_connection_lifetime", false)
			viper.SetDefault("project_watch", false)
			viper.SetDefault("project_publish", false)
			viper.SetDefault("project_anonymous", false)
			viper.SetDefault("project_presence", false)
			viper.SetDefault("project_history_size", 0)
			viper.SetDefault("project_history_lifetime", 0)
			viper.SetDefault("project_namespaces", "")

			// Tarantool defaults
			/*
				viper.SetDefault("tt_pool", 2)
				viper.SetDefault("tt_host", "127.0.0.1")
				viper.SetDefault("tt_port", "3301")
				viper.SetDefault("tt_user", "")
				viper.SetDefault("tt_password", "")
				viper.SetDefault("tt_timeout_response", 500)
				viper.SetDefault("tt_timeout_reconnect", 500)
			*/

			viper.SetEnvPrefix("centrifugo")
			viper.BindEnv("engine")
			viper.BindEnv("insecure")
			viper.BindEnv("web_password")
			viper.BindEnv("web_secret")
			viper.BindEnv("project_name")
			viper.BindEnv("project_secret")
			viper.BindEnv("project_connection_lifetime")
			viper.BindEnv("project_watch")
			viper.BindEnv("project_publish")
			viper.BindEnv("project_anonymous")
			viper.BindEnv("project_join_leave")
			viper.BindEnv("project_presence")
			viper.BindEnv("project_history_size")
			viper.BindEnv("project_history_lifetime")

			viper.BindPFlag("port", cmd.Flags().Lookup("port"))
			viper.BindPFlag("address", cmd.Flags().Lookup("address"))
			viper.BindPFlag("debug", cmd.Flags().Lookup("debug"))
			viper.BindPFlag("name", cmd.Flags().Lookup("name"))
			viper.BindPFlag("web", cmd.Flags().Lookup("web"))
			viper.BindPFlag("engine", cmd.Flags().Lookup("engine"))
			viper.BindPFlag("insecure", cmd.Flags().Lookup("insecure"))
			viper.BindPFlag("ssl", cmd.Flags().Lookup("ssl"))
			viper.BindPFlag("ssl_cert", cmd.Flags().Lookup("ssl_cert"))
			viper.BindPFlag("ssl_key", cmd.Flags().Lookup("ssl_key"))
			viper.BindPFlag("log_level", cmd.Flags().Lookup("log_level"))
			viper.BindPFlag("log_file", cmd.Flags().Lookup("log_file"))
			viper.BindPFlag("redis_host", cmd.Flags().Lookup("redis_host"))
			viper.BindPFlag("redis_port", cmd.Flags().Lookup("redis_port"))
			viper.BindPFlag("redis_password", cmd.Flags().Lookup("redis_password"))
			viper.BindPFlag("redis_db", cmd.Flags().Lookup("redis_db"))
			viper.BindPFlag("redis_url", cmd.Flags().Lookup("redis_url"))
			viper.BindPFlag("redis_api", cmd.Flags().Lookup("redis_api"))
			viper.BindPFlag("redis_pool", cmd.Flags().Lookup("redis_pool"))

			// Tarantool
			viper.BindPFlag("tt_pool", cmd.Flags().Lookup("tt_pool"))
			viper.BindPFlag("tt_host", cmd.Flags().Lookup("tt_host"))
			viper.BindPFlag("tt_port", cmd.Flags().Lookup("tt_port"))
			viper.BindPFlag("tt_user", cmd.Flags().Lookup("tt_user"))
			viper.BindPFlag("tt_password", cmd.Flags().Lookup("tt_password"))
			viper.BindPFlag("tt_timeout_response", cmd.Flags().Lookup("tt_timeout_request"))
			viper.BindPFlag("tt_timeout_reconnect", cmd.Flags().Lookup("tt_timeout_reconnect"))
			viper.BindPFlag("tt_max_reconnect", cmd.Flags().Lookup("tt_max_reconnect"))
			viper.BindPFlag("tt_aggr_node", cmd.Flags().Lookup("tt_aggr_node"))
			viper.BindPFlag("tt_aggr_port", cmd.Flags().Lookup("tt_aggr_port"))
			viper.BindPFlag("tt_service_name", cmd.Flags().Lookup("tt_service_name"))
			viper.BindPFlag("tt_acl", cmd.Flags().Lookup("tt_acl"))

			err := validateConfig(configFile)
			if err != nil {
				logger.FATAL.Fatalln(err)
			}

			viper.SetConfigFile(configFile)
			err = viper.ReadInConfig()
			if err != nil {
				logger.FATAL.Fatalln("Unable to locate config file")
			}
			setupLogging()

			if os.Getenv("GOMAXPROCS") == "" {
				if viper.IsSet("gomaxprocs") && viper.GetInt("gomaxprocs") > 0 {
					runtime.GOMAXPROCS(viper.GetInt("gomaxprocs"))
				} else {
					runtime.GOMAXPROCS(runtime.NumCPU())
				}
			}

			logger.INFO.Println("GOMAXPROCS set to", runtime.GOMAXPROCS(0))
			logger.INFO.Println("Using config file:", viper.ConfigFileUsed())

			c := newConfig()
			s := structureFromConfig(nil)

			app, err := libcentrifugo.NewApplication(c)
			if err != nil {
				logger.FATAL.Fatalln(err)
			}

			app.SetStructure(s)

			var e libcentrifugo.Engine
			switch viper.GetString("engine") {
			case "memory":
				e = libcentrifugo.NewMemoryEngine(app)
			case "redis":
				e = libcentrifugo.NewRedisEngine(
					app,
					viper.GetString("redis_host"),
					viper.GetString("redis_port"),
					viper.GetString("redis_password"),
					viper.GetString("redis_db"),
					viper.GetString("redis_url"),
					viper.GetBool("redis_api"),
					viper.GetInt("redis_pool"),
				)
			case "tarantool":
				hostname, err := os.Hostname()
				if err != nil {
					logger.FATAL.Fatalln("Unable to determine local hostname: " + err.Error())
				}
				config := libcentrifugo.TarantoolEngineConfig{
					PoolConfig: libcentrifugo.TarantoolPoolConfig{
						Address:  viper.GetString("tt_host") + ":" + viper.GetString("tt_port"),
						PoolSize: viper.GetInt("tt_pool"),
						Opts: tarantool.Opts{
							Timeout: time.Duration(viper.GetInt("tt_timeout_response")) * time.Millisecond,
							Reconnect: time.Duration(viper.GetInt("tt_timeout_reconnect")) * time.Millisecond,
							MaxReconnects: viper.GetSizeInBytes("tt_max_reconnect"),
							User: viper.GetString("tt_user"),
							Pass: viper.GetString("tt_password"),
						},
					},
					TTConnector: extender.Config{
						AggrNode:    viper.GetString("tt_aggr_node"),
						AggrPort:    viper.GetString("tt_aggr_port"),
						ServiceName: viper.GetString("tt_service_name"),
						ACL:         viper.GetString("tt_acl"),
					},
					Endpoint: fmt.Sprintf("http://%s:%s", hostname, viper.GetString("port")),
				}
				e = libcentrifugo.NewTarantoolEngine(app, config)
			default:
				logger.FATAL.Fatalln("unknown engine: " + viper.GetString("engine"))
			}

			logger.INFO.Println("Engine:", viper.GetString("engine"))
			logger.DEBUG.Printf("%v\n", viper.AllSettings())
			logger.INFO.Println("Use SSL:", viper.GetBool("ssl"))
			if viper.GetBool("ssl") {
				if viper.GetString("ssl_cert") == "" {
					logger.FATAL.Println("No SSL certificate provided")
					os.Exit(1)
				}
				if viper.GetString("ssl_key") == "" {
					logger.FATAL.Println("No SSL certificate key provided")
					os.Exit(1)
				}
			}
			app.SetEngine(e)

			app.Run()

			go handleSignals(app)

			mux := libcentrifugo.DefaultMux(app, viper.GetString("prefix"), viper.GetString("web"), viper.GetString("sockjs_url"))

			addr := viper.GetString("address") + ":" + viper.GetString("port")
			logger.INFO.Printf("Start serving on %s\n", addr)
			if useSSL {
				if err := http.ListenAndServeTLS(addr, sslCert, sslKey, mux); err != nil {
					logger.FATAL.Fatalln("ListenAndServe:", err)
				}
			} else {
				if err := http.ListenAndServe(addr, mux); err != nil {
					logger.FATAL.Fatalln("ListenAndServe:", err)
				}
			}
		},
	}
	rootCmd.Flags().StringVarP(&port, "port", "p", "8000", "port to bind to")
	rootCmd.Flags().StringVarP(&address, "address", "a", "", "address to listen on")
	rootCmd.Flags().BoolVarP(&debug, "debug", "d", false, "debug mode - please, do not use it in production")
	rootCmd.Flags().StringVarP(&configFile, "config", "c", "config.json", "path to config file")
	rootCmd.Flags().StringVarP(&name, "name", "n", "", "unique node name")
	rootCmd.Flags().StringVarP(&web, "web", "w", "", "optional path to web interface application")
	rootCmd.Flags().StringVarP(&engn, "engine", "e", "memory", "engine to use: memory, redis, tarantool")
	rootCmd.Flags().BoolVarP(&insecure, "insecure", "", false, "start in insecure mode")
	rootCmd.Flags().BoolVarP(&useSSL, "ssl", "", false, "accept SSL connections. This requires an X509 certificate and a key file")
	rootCmd.Flags().StringVarP(&sslCert, "ssl_cert", "", "", "path to an X509 certificate file")
	rootCmd.Flags().StringVarP(&sslKey, "ssl_key", "", "", "path to an X509 certificate key")
	rootCmd.Flags().StringVarP(&logLevel, "log_level", "", "info", "set the log level: debug, info, error, critical, fatal or none")
	rootCmd.Flags().StringVarP(&logFile, "log_file", "", "", "optional log file - if not specified all logs go to STDOUT")
	rootCmd.Flags().StringVarP(&redisHost, "redis_host", "", "127.0.0.1", "redis host (Redis engine)")
	rootCmd.Flags().StringVarP(&redisPort, "redis_port", "", "6379", "redis port (Redis engine)")
	rootCmd.Flags().StringVarP(&redisPassword, "redis_password", "", "", "redis auth password (Redis engine)")
	rootCmd.Flags().StringVarP(&redisDB, "redis_db", "", "0", "redis database (Redis engine)")
	rootCmd.Flags().StringVarP(&redisURL, "redis_url", "", "", "redis connection URL (Redis engine)")
	rootCmd.Flags().BoolVarP(&redisAPI, "redis_api", "", false, "enable Redis API listener (Redis engine)")
	rootCmd.Flags().IntVarP(&redisPool, "redis_pool", "", 256, "Redis pool size (Redis engine)")

	// Tarantool
	rootCmd.Flags().StringVarP(&ttHost, "tt_host", "", "127.0.0.1", "tarantool host (Tarantool engine)")
	rootCmd.Flags().StringVarP(&ttPort, "tt_port", "", "3301", "tarantool port (Tarantool engine)")
	rootCmd.Flags().StringVarP(&ttUser, "tt_user", "", "", "tarantool user (Tarantool engine)")
	rootCmd.Flags().StringVarP(&ttPassword, "tt_password", "", "", "tarantool password (Tarantool engine)")
	rootCmd.Flags().IntVarP(&ttPool, "tt_pool", "", 2, "tarantool connection pool size (Tarantool engine)")
	rootCmd.Flags().IntVarP(&ttTimeoutResponse, "tt_timeout_response", "", 500, "timeout to wait response in milliseconds (Tarantool engine)")
	rootCmd.Flags().IntVarP(&ttTimeoutReconnect, "tt_timeout_reconnect", "", 500, "timeout to wait until reconnection attempt in milliseconds (Tarantool engine)")
	rootCmd.Flags().UintVarP(&ttMaxReconnect, "tt_max_reconnect", "", 1<<32-1, "max number of reconnection attempts (Tarantool engine)") // uint32 max
	rootCmd.Flags().StringVarP(&ttAggrNode, "tt_aggr_node", "", "localhost", "cocaine aggregation node (Tarantool engine)")
	rootCmd.Flags().StringVarP(&ttAggrPort, "tt_aggr_port", "", "10053", "cocaine aggregation node port (Tarantool engine)")
	rootCmd.Flags().StringVarP(&ttServiceName, "tt_service_name", "", "ttconnector", "cocaine TTConnector service name (Tarantool engine)")
	rootCmd.Flags().StringVarP(&ttACL, "tt_acl", "", "test_acl", "TTConnector ACL (Tarantool engine)")

	var versionCmd = &cobra.Command{
		Use:   "version",
		Short: "Centrifugo version number",
		Long:  `Print the version number of Centrifugo`,
		Run: func(cmd *cobra.Command, args []string) {
			fmt.Printf("Centrifugo v%s\n", VERSION)
		},
	}

	var checkConfigFile string

	var checkConfigCmd = &cobra.Command{
		Use:   "checkconfig",
		Short: "Check configuration file",
		Long:  `Check Centrifugo configuration file`,
		Run: func(cmd *cobra.Command, args []string) {
			err := validateConfig(checkConfigFile)
			if err != nil {
				logger.FATAL.Fatalln(err)
			}
		},
	}
	checkConfigCmd.Flags().StringVarP(&checkConfigFile, "config", "c", "config.json", "path to config file to check")

	var outputConfigFile string

	var generateConfigCmd = &cobra.Command{
		Use:   "genconfig",
		Short: "Generate simple configuration file to start with",
		Long:  `Generate simple configuration file to start with`,
		Run: func(cmd *cobra.Command, args []string) {
			err := generateConfig(outputConfigFile)
			if err != nil {
				logger.FATAL.Fatalln(err)
			}
		},
	}
	generateConfigCmd.Flags().StringVarP(&outputConfigFile, "config", "c", "config.json", "path to output config file")

	rootCmd.AddCommand(versionCmd)
	rootCmd.AddCommand(checkConfigCmd)
	rootCmd.AddCommand(generateConfigCmd)
	rootCmd.Execute()
}

func main() {
	Main()
}
